File: src/main/java/org/example/DatabaseUtilities.java
package org.example;

import com.j256.ormlite.dao.Dao;
import com.j256.ormlite.dao.DaoManager;
import com.j256.ormlite.jdbc.JdbcPooledConnectionSource;
import com.j256.ormlite.table.TableUtils;
import org.example.data.*;

import java.sql.SQLException;

public class DatabaseUtilities {
    private static final int VERSION = 1;
    private static JdbcPooledConnectionSource connectionSource;
    private static Dao<User, Integer> userDao;
    private static Dao<Authentication, Integer> authenticationDao;
    private static Dao<Channel, Integer> channelDao;
    private static Dao<Entry, Integer> entryDao;
    private static Dao<UserChannelLink, Integer> userChannelLinkDao;
    private static Dao<Filter, Integer> filterDao;
    private static Dao<FilterChannelLink, Integer> filterChannelLinkDao;
    private static Dao<FilterEntryLink, Integer> filterEntryLinkDao;
    public static void init() {
        Config config = Config.getConfig();
        int tries = 0;
        boolean success = false;
        while (tries<config.maxRetries) {
            try {
                connectionSource = new JdbcPooledConnectionSource("jdbc:mariadb://" + config.dbUrl + ":" + config.dbPort + "/" + config.dbDatabase, config.dbUser, config.dbPass);
                connectionSource.setMaxConnectionsFree(config.maxConnections);
                connectionSource.initialize();
                success = true;
                break;
            } catch (SQLException e) {
                tries++;
                try {
                    Thread.sleep(config.retryDelay);
                } catch (InterruptedException ignored) {
                }
            }
        }
        if (!success) {
            throw new RuntimeException("Failed to connect to the database");
        }
        tries = 0;
        while (tries<config.maxRetries) {
            try {
                TableUtils.createTableIfNotExists(connectionSource, User.class);
                userDao = DaoManager.createDao(connectionSource, User.class);
                userDao.setObjectCache(true);
                break;
            } catch (SQLException e) {
                tries++;
                try {
                    Thread.sleep(config.retryDelay);
                } catch (InterruptedException ignored) {
                }
            }
        }
        if (userDao==null) {
            throw new RuntimeException("Failed to create DAO for User");
        }
        tries = 0;
        while (tries<config.maxRetries) {
            try {
                TableUtils.createTableIfNotExists(connectionSource, Authentication.class);
                authenticationDao = DaoManager.createDao(connectionSource, Authentication.class);
                authenticationDao.setObjectCache(true);
                break;
            } catch (SQLException e) {
                tries++;
                try {
                    Thread.sleep(config.retryDelay);
                } catch (InterruptedException ignored) {
                }
            }
        }
        if (authenticationDao==null) {
            throw new RuntimeException("Failed to create DAO for Authentication");
        }
        tries = 0;
        while (tries<config.maxRetries) {
            try {
                TableUtils.createTableIfNotExists(connectionSource, Channel.class);
                channelDao = DaoManager.createDao(connectionSource, Channel.class);
                channelDao.setObjectCache(true);
                break;
            } catch (SQLException e) {
                tries++;
                try {
                    Thread.sleep(config.retryDelay);
                } catch (InterruptedException ignored) {
                }
            }
        }
        if (channelDao==null) {
            throw new RuntimeException("Failed to create DAO for Channel");
        }
        tries = 0;
        while (tries<config.maxRetries) {
            try {
                TableUtils.createTableIfNotExists(connectionSource, Entry.class);
                entryDao = DaoManager.createDao(connectionSource, Entry.class);
                entryDao.setObjectCache(true);
                break;
            } catch (SQLException e) {
                tries++;
                try {
                    Thread.sleep(config.retryDelay);
                } catch (InterruptedException ignored) {
                }
            }
        }
        if (entryDao==null) {
            throw new RuntimeException("Failed to create DAO for Entry");
        }
        tries = 0;
        while (tries<config.maxRetries) {
            try {
                TableUtils.createTableIfNotExists(connectionSource, UserChannelLink.class);
                userChannelLinkDao = DaoManager.createDao(connectionSource, UserChannelLink.class);
                userChannelLinkDao.setObjectCache(true);
                break;
            } catch (SQLException e) {
                tries++;
                try {
                    Thread.sleep(config.retryDelay);
                } catch (InterruptedException ignored) {
                }
            }
            tries = 0;
            while (tries<config.maxRetries) {
                try {
                    TableUtils.createTableIfNotExists(connectionSource, Filter.class);
                    filterDao = DaoManager.createDao(connectionSource, Filter.class);
                    filterDao.setObjectCache(true);
                    break;
                } catch (SQLException e) {
                    tries++;
                    try {
                        Thread.sleep(config.retryDelay);
                    } catch (InterruptedException ignored) {
                    }
                }
            }
            if (filterDao==null) {
                throw new RuntimeException("Failed to create DAO for Filter");
            }
            tries = 0;
            while (tries<config.maxRetries) {
                try {
                    TableUtils.createTableIfNotExists(connectionSource, FilterChannelLink.class);
                    filterChannelLinkDao = DaoManager.createDao(connectionSource, FilterChannelLink.class);
                    filterChannelLinkDao.setObjectCache(true);
                    break;
                } catch (SQLException e) {
                    tries++;
                    try {
                        Thread.sleep(config.retryDelay);
                    } catch (InterruptedException ignored) {
                    }
                }
            }
            if (filterChannelLinkDao==null) {
                throw new RuntimeException("Failed to create DAO for FilterChannelLink");
            }
            tries = 0;
            while (tries<config.maxRetries) {
                try {
                    TableUtils.createTableIfNotExists(connectionSource, FilterEntryLink.class);
                    filterEntryLinkDao = DaoManager.createDao(connectionSource, FilterEntryLink.class);
                    filterEntryLinkDao.setObjectCache(true);
                    break;
                } catch (SQLException e) {
                    tries++;
                    try {
                        Thread.sleep(config.retryDelay);
                    } catch (InterruptedException ignored) {
                    }
                }
            }
            if (filterEntryLinkDao==null) {
                throw new RuntimeException("Failed to create DAO for FilterEntryLink");
            }
        }
        if (userChannelLinkDao==null) {
            throw new RuntimeException("Failed to create DAO for UserChannelLink");
        }
        tries = 0;
        while (tries<config.maxRetries) {
            try {
                TableUtils.createTableIfNotExists(connectionSource, Filter.class);
                filterDao = DaoManager.createDao(connectionSource, Filter.class);
                filterDao.setObjectCache(true);
                break;
            } catch (SQLException e) {
                tries++;
                try {
                    Thread.sleep(config.retryDelay);
                } catch (InterruptedException ignored) {
                }
            }
        }
        if (filterDao==null) {
            throw new RuntimeException("Failed to create DAO for Filter");
        }
        tries = 0;
        while (tries<config.maxRetries) {
            try {
                TableUtils.createTableIfNotExists(connectionSource, FilterChannelLink.class);
                filterChannelLinkDao = DaoManager.createDao(connectionSource, FilterChannelLink.class);
                filterChannelLinkDao.setObjectCache(true);
                break;
            } catch (SQLException e) {
                tries++;
                try {
                    Thread.sleep(config.retryDelay);
                } catch (InterruptedException ignored) {
                }
            }
        }
        if (filterChannelLinkDao==null) {
            throw new RuntimeException("Failed to create DAO for FilterChannelLink");
        }
        tries = 0;
        while (tries<config.maxRetries) {
            try {
                TableUtils.createTableIfNotExists(connectionSource, FilterEntryLink.class);
                filterEntryLinkDao = DaoManager.createDao(connectionSource, FilterEntryLink.class);
                filterEntryLinkDao.setObjectCache(true);
                break;
            } catch (SQLException e) {
                tries++;
                try {
                    Thread.sleep(config.retryDelay);
                } catch (InterruptedException ignored) {
                }
            }
        }
        if (filterEntryLinkDao==null) {
            throw new RuntimeException("Failed to create DAO for FilterEntryLink");
        }
    }
    public static void upgrade() {
        Config config = Config.getConfig();
        if (config.version>VERSION) {
            throw new RuntimeException("Database version is too high");
        }
        switch (config.version) {
            case 1:
                // Nothing to upgrade yet as we are on the first version
        }
    }

    public static Dao<User, Integer> getUserDao() {
        return userDao;
    }
    public static Dao<Authentication, Integer> getAuthenticationDao() {
        return authenticationDao;
    }
    public static Dao<Channel, Integer> getChannelDao() {
        return channelDao;
    }
    public static Dao<Entry, Integer> getEntryDao() {
        return entryDao;
    }
    public static Dao<UserChannelLink, Integer> getUserChannelLinkDao() {
        return userChannelLinkDao;
    }
    public static Dao<Filter, Integer> getFilterDao() {
        return filterDao;
    }
    public static Dao<FilterChannelLink, Integer> getFilterChannelLinkDao() {
        return filterChannelLinkDao;
    }
    public static Dao<FilterEntryLink, Integer> getFilterEntryLinkDao() {
        return filterEntryLinkDao;
    }
}

File: src/main/kotlin/DatabaseUtilities.kt
package org.example

import com.j256.ormlite.dao.Dao
import com.j256.ormlite.dao.DaoManager
import com.j256.ormlite.jdbc.JdbcPooledConnectionSource
import com.j256.ormlite.table.TableUtils
import org.example.data.*
import java.sql.SQLException

private val config: Config = Config.getConfig()
private const val version: Int = 1
private var connectionSource: JdbcPooledConnectionSource? = null
lateinit var userDao: Dao<User, Int>
lateinit var channelDao: Dao<Channel, Int>
lateinit var entryDao: Dao<Entry, Int>
lateinit var authenticationDao: Dao<Authentication, Int>
lateinit var filterDao: Dao<Filter, Int>
lateinit var filterChannelLinkDao: Dao<FilterChannelLink, Int>
lateinit var filterEntryLinkDao: Dao<FilterEntryLink, Int>
lateinit var userChannelLinkDao: Dao<UserChannelLink, Int>
fun initDatabase() {
    var tries: Int = 0
    var success: Boolean = false
    while (!success) {
        try {
            connectionSource = JdbcPooledConnectionSource("jdbc:mariadb://" + config.databaseSettings.databaseURL + ":" + config.databaseSettings.databasePort + "/" + config.databaseSettings.databaseName, config.databaseSettings.databaseUser, config.databaseSettings.databasePassword)
            connectionSource?.setMaxConnectionsFree(config.databaseSettings.databaseMaxConnections)
            connectionSource?.initialize()
            TableUtils.createTableIfNotExists(connectionSource, User::class.java)
            userDao = DaoManager.createDao(connectionSource, User::class.java)
            userDao.setObjectCache(true)
            TableUtils.createTableIfNotExists(connectionSource, Channel::class.java)
            channelDao = DaoManager.createDao(connectionSource, Channel::class.java)
            channelDao.setObjectCache(true)
            TableUtils.createTableIfNotExists(connectionSource, Entry::class.java)
            entryDao = DaoManager.createDao(connectionSource, Entry::class.java)
            entryDao.setObjectCache(true)
            TableUtils.createTableIfNotExists(connectionSource, Authentication::class.java)
            authenticationDao = DaoManager.createDao(connectionSource, Authentication::class.java)
            authenticationDao.setObjectCache(true)
            TableUtils.createTableIfNotExists(connectionSource, Filter::class.java)
            filterDao = DaoManager.createDao(connectionSource, Filter::class.java)
            filterDao.setObjectCache(true)
            TableUtils.createTableIfNotExists(connectionSource, FilterChannelLink::class.java)
            filterChannelLinkDao = DaoManager.createDao(connectionSource, FilterChannelLink::class.java)
            filterChannelLinkDao.setObjectCache(true)
            TableUtils.createTableIfNotExists(connectionSource, FilterEntryLink::class.java)
            filterEntryLinkDao = DaoManager.createDao(connectionSource, FilterEntryLink::class.java)
            filterEntryLinkDao.setObjectCache(true)
            TableUtils.createTableIfNotExists(connectionSource, UserChannelLink::class.java)
            userChannelLinkDao = DaoManager.createDao(connectionSource, UserChannelLink::class.java)
            userChannelLinkDao.setObjectCache(true)
            success = true
        } catch (e: SQLException) {
            tries++
            if (tries >= config.databaseSettings.databaseMaxRetries) {
                throw e
            }
            Thread.sleep(config.databaseSettings.databaseRetryDelay.toLong())
        }
    }
}
fun upgrade() {
    if (config.version > version) {
        throw RuntimeException("Database version is newer than application version")
    }
    when(config.version) {
        1-> println("Nothing to do for upgrade")
    }
}